src/com/facebook/buck/go/GoCompileStep.java

```diff
@@ -69,37 +70,45 @@ public GoCompileStep(
 
   @Override
   protected ImmutableList<String> getShellCommandInternal(ExecutionContext context) {
-    ImmutableList.Builder<String> commandBuilder =
-        ImmutableList.<String>builder()
-            .addAll(compilerCommandPrefix)
-            .add("-p", packageName.toString())
-            .add("-pack")
-            .add("-trimpath", workingDirectory.toString())
-            .add("-nolocalimports")
-            .addAll(flags)
-            .add("-o", output.toString());
-
-    for (Path dir : includeDirectories) {
-      commandBuilder.add("-I", dir.toString());
+    ArrayList<String> pathStrings = new ArrayList<>();
+    for (Path path : srcs) {
+      pathStrings.add(path.toString());
     }
+    if (pathStrings.size() > 0) {
+      ImmutableList.Builder<String> commandBuilder =
+          ImmutableList.<String>builder()
+              .addAll(compilerCommandPrefix)
+              .add("-p", packageName.toString())
+              .add("-pack")
+              .add("-trimpath", workingDirectory.toString())
+              .add("-nolocalimports")
+              .addAll(flags)
+              .add("-o", output.toString());
 
-    for (Map.Entry<Path, Path> importMap : importPathMap.entrySet()) {
-      commandBuilder.add("-importmap", importMap.getKey() + "=" + importMap.getValue());
-    }
+      for (Path dir : includeDirectories) {
+        commandBuilder.add("-I", dir.toString());
+      }
 
-    if (asmHeaderPath.isPresent()) {
-      commandBuilder.add("-asmhdr", asmHeaderPath.get().toString());
-    }
+      for (Map.Entry<Path, Path> importMap : importPathMap.entrySet()) {
+        commandBuilder.add("-importmap", importMap.getKey() + "=" + importMap.getValue());
+      }
 
-    if (!allowExternalReferences) {
-      // -complete means the package does not use any non Go code, so external functions
-      // (e.g. Cgo, asm) aren't allowed.
-      commandBuilder.add("-complete");
-    }
+      if (asmHeaderPath.isPresent()) {
+        commandBuilder.add("-asmhdr", asmHeaderPath.get().toString());
+      }
+
+      if (!allowExternalReferences) {
+        // -complete means the package does not use any non Go code, so external functions
+        // (e.g. Cgo, asm) aren't allowed.
+        commandBuilder.add("-complete");
+      }
 
-    commandBuilder.addAll(srcs.stream().map(Object::toString).iterator());
+      commandBuilder.addAll(pathStrings);
 
-    return commandBuilder.build();
+      return commandBuilder.build();
+    } else {
+      return ImmutableList.of("echo", "No source files found in " + workingDirectory);
```

Comment:
Why do you need to call `echo`?

Reply:
The caller of this method will throw an exception if I return an empty list. So I have to return some command. What other command do you suggest?

Reply:
Why not throw an exception?

Reply:
Throwing an exception here will crash buck. As this step doesn't expect any stdout, putting a message in the stdout will result in a warning message in Buck's output (colored as red). If there are things in this package needed by other packages, the build will fail with message saying something is undefined. Combined with the warning message here, it's easy to track down the problem. If the build succeeds regardless, it means this package is not necessary for the current platform, although it might be necessary for other platforms. Buck should tolerate such scenario too.

Reply:
The compile command will fail if there is no source file, producing an error message prompting the correct format of the command line arguments, which can be used in place of this echo message. I do the due diligence check for lack of input files [here](https://github.com/facebook/buck/pull/1782/files#diff-d4c3760f594f832155f3915c72110e06R171), but the existence of raw source files doesn't guarantee the existence of filtered source files.

Reply:
Then if srcs is empty we can skip this step, right?

Reply:
The length of `srcs` can not be determined until GoListStep is executed. So in GoCompile.getBuildSteps, we are not always sure whether to skip GoCompileStep. In this method, when GoCompileStep is about to execute, it's probably too late to yank it out. Do you see a way to skip GoCompileStep after GoListStep is executed?

Reply:
This issue shows the problem with supporting the directories as inputs in general. Buck need to know about the inputs before creating these steps, we shouldn't skip steps based on the output from the previous steps.

Going to back to the original issue of supporting directories as inputs in go rules, Java, for example, supports using zip files as inputs. This gives you a way to pass multiple files as a single file. Is there a reason we cannot add support for that in go rules?

Reply:
This issue is really not about supporting directories. The FilteredSourceFiles only takes individual source files, not directories. What it does is run a "go list" command on the parent directory of each file, and decide whether to ignore the source file. Removing the support for directories doesn't affect any code in this filtering logic. One can still pass a Windows specific Go file into go_library, by either using glob or the file name, and try to compile it on Linux. This filtering still has to happen in order to have the compiler skip that file.

Reply:
Is running this "go list" command a prerequisite for a build or it's just for the sake of convenience?

The issue as I see it is the steps generated in `FilteredSourceFiles`are accessing the whole directory and it's not clear whether the files in that directory are dependencies of the build rule that produced the steps or not.

It still feels to me that this logic is more complicated than necessary because of the support of directories.

Reply:
Go source files have [build constraints](https://golang.org/pkg/go/build/#hdr-Build_Constraints), which specify whether a source file should be compiled for the current platform. In order for Buck to honor the build constraints, there are two options:

1. Running "go list" command on the directory each go file resides. As "go list" command only returns source files with all build constraints satisfied, we can know whether we should pass a file to the compiler by checking if it's in the output of "go list"

2. Implement a constraints solver ourselves. Then it won't rely on "go list" command, and we don't need GoListStep anymore. Also, we will know early on whether to skip GoCompileStep in GoCompile.getBuildSteps. However, the constraint solver will need to update whenever the definition of Go build constraints changes

In either option, the build constraints are checked on file-to-file basis. All directories passed in from build rules have been expanded before the filtering logic takes place, so the filtering logic only sees files, not directories. Do you see how this logic can be simplified if the directory support is removed?

Reply:
How about overriding `execute` to check if `srcs` is empty, return `StepExecutionResult.of(0)` if it is and proceed with compilation if some source files are present?

Reply:
Good idea. Updated now